File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int l = temperatures.size();
        vector<int>answer(l,0);
        stack<pair<int,int>>s;
        pair<int,int>p;
        
        for(int i=0;i<l;i++){
            if(s.empty()){
                s.push({temperatures[i],i});
            }
            else{
                p = s.top();
                while(s.size()!= 0 && (temperatures[i] > p.first)){
                    answer[p.second] = i - p.second;
                    s.pop();
                    if(s.size()!= 0){
                        p = s.top();
                    }
                }
                s.push({temperatures[i],i});
            }
        }
        return answer;
    }
};

File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int hottestTemp = 0;
        int n = temperatures.length;
        int[] res = new int[n];

        for(int i = n-1; i>=0; i--){
            int currentTemp = temperatures[i];
            if(currentTemp >= hottestTemp){
                hottestTemp = currentTemp;
                continue;
            }

            int days = 1;
            while(temperatures[i + days] <= currentTemp){
                days += res[i + days];
            }
            res[i] = days;
        }
        return res;
    }
}